puffin/src/scope_details.rs

```diff
@@ -0,0 +1,239 @@
+        match self {
+            ScopeType::Function => "function scope",
+            ScopeType::Scope => "scope",
+        }
+    }
+}
+
+impl ScopeType {
+    pub fn function_scope() -> Self {
+        Self::Function
+    }
+
+    pub fn scope() -> Self {
+        Self::Scope
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+#[cfg_attr(
+    feature = "serialization",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+/// Detailed information about a scope.
+pub struct ScopeDetails {
+    /// Unique scope identifier.
+    /// Always initialized once registered.
+    /// It is `None` when an external library has yet to register this scope.
+    pub(crate) scope_id: Option<ScopeId>,
+    /// A name for a profile scope, a function profile scope does not have a custom provided name.
+    pub scope_name: Option<Cow<'static, str>>,
+    /// The function name of the function in which this scope is contained.
+    /// The name might be slightly modified to represent a short descriptive representation.
+    pub function_name: Cow<'static, str>,
+    /// The file path in which this scope is contained.
+    /// The path might be slightly modified to represent a short descriptive representation.
+    pub file_path: Cow<'static, str>,
+    /// The exact line number at which this scope is located.
+    pub line_nr: u32,
+}
+
+impl ScopeDetails {
+    /// Creates a new user scope with a unique name.
+    pub fn from_scope_name<T>(scope_name: T) -> Self
+    where
+        T: Into<Cow<'static, str>>,
+    {
+        Self {
+            scope_id: None,
+            scope_name: Some(scope_name.into()),
+            function_name: Default::default(),
+            file_path: Default::default(),
+            line_nr: Default::default(),
+        }
+    }
+
+    /// Creates a new user scope with a unique id allocated by puffin.
+    /// This function should not be exposed as only puffin should allocate ids for scopes.
+    pub(crate) fn from_scope_id(scope_id: ScopeId) -> Self {
+        Self {
+            scope_id: Some(scope_id),
+            scope_name: None,
+            function_name: Default::default(),
+            file_path: Default::default(),
+            line_nr: Default::default(),
+        }
+    }
+
+    #[inline]
+    pub fn with_function_name<T>(mut self, name: T) -> Self
+    where
+        T: Into<Cow<'static, str>>,
+    {
+        self.function_name = name.into();
+        self
+    }
+
+    #[inline]
+    pub fn with_file<T>(mut self, file: T) -> Self
+    where
+        T: Into<Cow<'static, str>>,
+    {
+        self.file_path = file.into();
+        self
+    }
+
+    #[inline]
+    pub fn with_line_nr(mut self, line_nr: u32) -> Self {
+        self.line_nr = line_nr;
+        self
+    }
+
+    /// Returns the scope name if this is a profile scope or else the function name.
+    pub fn name(&self) -> &Cow<'static, str> {
+        self.scope_name.as_ref().map_or(&self.function_name, |x| x)
```

Comment:
```suggestion
    #[inline]
    pub fn name(&self) -> &Cow<'static, str> {
        self.scope_name.as_ref().map_or(&self.function_name, |x| x)
```

possibly this could return a `&str` instead for simplicity